rfc_samplers.md

```diff
@@ -0,0 +1,107 @@
+**DISCLAIMER** This RFC is currently lacking (at least):
+
+- a "time-based" equivalent of dilation
+- a "time-based" equivalent of random uniform sampling.
+
+Main design principles
+----------------------
+
+- The existing feature space of different libraries (torchvision, torchmm,
+  internal stuff) is supported. In particular "Random", "Uniform" and
+  "Periodic" samplers are supported, with hopefully more descriptive names.
+- Sampler is agnostic to decoding options: the decoder object is passed to the
+  sampler by the user.
+- Explicit **non goal**: to support arbitrary strategy that we haven't
+  observed or heard usage of. For example, in all existing libraries the clip
+  content is never random and just determined by 2 parameters:
+  `num_frames_per_clip` and `step_between_frames` (dilation). We allow for
+  future extensibility, but we're not explicitly enabling additional
+  clip-content strategies.
+- The term "clip" is being used to denote a sequence of frames in presentation
+  order. The frames aren't necessarily consecutive (when
+  `step_between_frames > 1`). We should add this term to the glossary. I don't
+  feel too strongly about using the name "clip", but it's used in all existing
+  libraries, and I don't know of an alternative that would be universally
+  understood.
+
+
+Option 1
+--------
+
+- One function per sampling strategy
+- Note: this is not 100% stateless, the decoder object is seeked so there are
+  side effects.
+
+```py
+from torchcodec.decoders import SimpleVideoDecoder
+from torchcodec import samplers
+
+video = "cool_vid"
+decoder = SimpleVideoDecoder(video)
+
+clips = samplers.get_uniformly_random_clips(
+    decoder,
+    num_clips=12,
+    # clip content params:
+    num_frames_per_clip=4,
+    step_between_frames=2,  # often called "dilation"
+    # sampler-specific params:
+    prioritize_keyframes=False,  # to implement "fast" sampling.
+                                 # Might need to be a separate function.
+)
+
+clips = samplers.get_evenly_spaced_clips(  # often called "Uniform"
+    decoder,
+    num_clips=12,
+    # clip content params:
+    num_frames_per_clip=4,
+    step_between_frames=2,
+)
+
+clips = samplers.get_evenly_timed_clips(  # often called "Periodic"
+    decoder,
+    num_clips_per_second=3,
+    # clip content params:
+    num_frames_per_clip=4,
+    step_between_frames=2,
+    # sampler-specific params:
+    clip_range_start_seconds=3,
+    clip_range_end_seconds=4,
+)
```

Comment:
I do think @ahmadsharif1's point about variable FPS is relevant here. Clearly, `get_event_timed_clips()` is all about PTS as "time" is in the name. If we respect PTS for it (and I think we should), that takes care of variable FPS videos.

However, `get_uniformly_random_clips()` and `get_evenly_spaced_clips()` is less clear. Do we want to be uniformly random or evenly spaced across all of the frames, or across the time duration of the video? I think both behaviors can be reasonable for a user to want. That may need to be a parameter option, or different functions entirely.

Reply:
The time-based equivalent of `get_evenly_spaced_clips()` is `get_evenly_timed_clips()`.

For the random one, I agree we need to think more. There is also the "step_betwee_frames" parameter which right now is only index-based. We may want it to be time-based as well, and some users may want it to be time-based on an index-based random sampler (and vice-versa).